src/rdkafka_telemetry_encode.c

```diff
@@ -0,0 +1,869 @@
+                                 void *const *arg) {
+        opentelemetry_proto_metrics_v1_ScopeMetrics *scope_metrics =
+            (opentelemetry_proto_metrics_v1_ScopeMetrics *)*arg;
+        if (!pb_encode_tag_for_field(stream, field))
+                return false;
+
+        return pb_encode_submessage(
+            stream, opentelemetry_proto_metrics_v1_ScopeMetrics_fields,
+            scope_metrics);
+}
+
+static bool encode_resource_metrics(pb_ostream_t *stream,
+                                    const pb_field_t *field,
+                                    void *const *arg) {
+        opentelemetry_proto_metrics_v1_ResourceMetrics *resource_metrics =
+            (opentelemetry_proto_metrics_v1_ResourceMetrics *)*arg;
+        if (!pb_encode_tag_for_field(stream, field))
+                return false;
+
+        return pb_encode_submessage(
+            stream, opentelemetry_proto_metrics_v1_ResourceMetrics_fields,
+            resource_metrics);
+}
+
+static bool encode_key_value(pb_ostream_t *stream,
+                             const pb_field_t *field,
+                             void *const *arg) {
+        if (!pb_encode_tag_for_field(stream, field))
+                return false;
+        opentelemetry_proto_common_v1_KeyValue *key_value =
+            (opentelemetry_proto_common_v1_KeyValue *)*arg;
+        return pb_encode_submessage(
+            stream, opentelemetry_proto_common_v1_KeyValue_fields, key_value);
+}
+
+static bool encode_key_values(pb_ostream_t *stream,
+                              const pb_field_t *field,
+                              void *const *arg) {
+        rd_kafka_telemetry_key_values_repeated_t *kv_arr =
+            (rd_kafka_telemetry_key_values_repeated_t *)*arg;
+        size_t i;
+
+        for (i = 0; i < kv_arr->count; i++) {
+
+                opentelemetry_proto_common_v1_KeyValue *kv =
+                    kv_arr->key_values[i];
+                if (!pb_encode_tag_for_field(stream, field))
+                        return false;
+
+                if (!pb_encode_submessage(
+                        stream, opentelemetry_proto_common_v1_KeyValue_fields,
+                        kv))
+                        return false;
+        }
+        return true;
+}
+
+static bool is_per_broker_metric(rd_kafka_t *rk, int metric_idx) {
+        if (rk->rk_type == RD_KAFKA_PRODUCER &&
+            (metric_idx ==
+                 RD_KAFKA_TELEMETRY_METRIC_PRODUCER_NODE_REQUEST_LATENCY_AVG ||
+             metric_idx ==
+                 RD_KAFKA_TELEMETRY_METRIC_PRODUCER_NODE_REQUEST_LATENCY_MAX)) {
+                return true;
+        }
+        if (rk->rk_type == RD_KAFKA_CONSUMER &&
+            (metric_idx ==
+                 RD_KAFKA_TELEMETRY_METRIC_CONSUMER_NODE_REQUEST_LATENCY_AVG ||
+             metric_idx ==
+                 RD_KAFKA_TELEMETRY_METRIC_CONSUMER_NODE_REQUEST_LATENCY_MAX)) {
+                return true;
+        }
+        return false;
```

Comment:
Updated and also fixed the bug at the initial loop where `total_metrics_count` is being calculated. We were passing the loop index instead of `metrics_to_encode[i]`